packages/remix-react/browser.tsx

```diff
@@ -270,6 +247,35 @@ export function RemixBrowser(_props: RemixBrowserProps): ReactElement {
         window.location,
         window.__remixContext.basename
       );
+
+      // Hard reload if the matches we rendered on the server aren't the matches
+      // we matched in the client, otherwise we'll try to hydrate without the
+      // right modules and throw a hydration error, which can put React into an
+      // infinite hydration loop when hydrating the full `<html>` document.
+      // This is usually the result of 2 rapid back/forward clicks from an
+      // external site into a Remix app, where we initially start the load for
+      // one URL and while the JS chunks are loading a second forward click moves
+      // us to a new URL.
+      let ssrMatches = window.__remixContext.ssrMatches;
+      let hasDifferentSSRMatches =
+        (initialMatches || []).length !== ssrMatches.length ||
+        !(initialMatches || []).every((m, i) => ssrMatches[i] === m.route.id);
+
+      if (hasDifferentSSRMatches && !window.__remixContext.isSpaMode) {
+        let ssr = ssrMatches.join(",");
+        let client = (initialMatches || []).map((m) => m.route.id).join(",");
+        let errorMsg =
+          `SSR Matches (${ssr}) do not match client matches (${client}) at ` +
+          `time of hydration , reloading page...`;
+        console.error(errorMsg);
+
+        window.location.reload();
```

Comment:
This window reload seems to be creating an infinite reload loop on Remix 2.11.0 on a 404 page. With or without a custom `ErrorBoundary`. See #9821.

It's hard to break the browser and get it into debug mode when it's doing this, but it seems that it's because `hasDifferentSSRMatches` is set to `true` because `(initialMatches || []).length !== ssrMatches.length` is true. `initialMatches` is `0`, because `matchRoutes` doesn't match a 404 to the `root` route? while `ssrMatches` is length `1` with the `root` route.